src/client/datascience/jupyter/kernels/cellExecutionQueue.ts

```diff
@@ -0,0 +1,183 @@
+            // If something goes wrong in execution of cells or one cell, then cancel the remaining cells.
+            await this.cancel();
+            this.completion.reject(ex);
+        }
+    }
+    private async executeQueuedCells() {
+        const notebook = await this.notebookPromise;
+        this.queueOfCellsToExecute.forEach((exec) => traceCellMessage(exec.cell, 'Ready to execute'));
+        while (this.queueOfCellsToExecute.length) {
+            // Take the first item from the queue, this way we maintain order of cell executions.
+            const cellToExecute = this.queueOfCellsToExecute[0];
+            if (!cellToExecute) {
+                continue;
+            }
+            traceCellMessage(cellToExecute.cell, 'Before Execute individual cell');
+            const executionResult = await this.executeIndividualCell(cellToExecute, notebook);
+            traceCellMessage(cellToExecute.cell, `After Execute individual cell ${executionResult}`);
+            // If a cell has failed the get out.
+            if (
+                this.cancelledOrCompletedWithErrors ||
+                executionResult === vscodeNotebookEnums.NotebookCellRunState.Error
+            ) {
+                this.cancelledOrCompletedWithErrors = true;
+                traceInfo(`Cancel all remaining cells ${this.cancelledOrCompletedWithErrors} || ${executionResult}`);
+                await this.cancel();
+                break;
+            }
+            this.onCellExecutionCompleted(cellToExecute);
+        }
+    }
+    private async executeIndividualCell(
+        cellExecution: CellExecution,
+        notebook: INotebook
+    ): Promise<NotebookCellRunState | undefined> {
+        traceCellMessage(cellExecution.cell, 'Push cell into queue for execution');
+        const chainedExecution = this.chainedCellExecutionPromise.then(async () => {
+            traceCellMessage(cellExecution.cell, 'Get cell from queue for execution');
+            // Start execution
+            await cellExecution.start(notebook);
+
+            // The result promise will resolve when it completes.
+            return cellExecution.result.finally(() =>
+                traceCellMessage(cellExecution.cell, 'Cell from queue completed execution')
+            );
+        });
+        this.chainedCellExecutionPromise = chainedExecution;
+        return chainedExecution;
+    }
+
+    private queueCellForExecution(cell: NotebookCell) {
+        const existingCellExecution = this.cellExecutions.get(cell);
+        if (existingCellExecution) {
+            traceCellMessage(cell, 'Use existing cell execution');
+            return existingCellExecution;
+        }
+        const cellExecution = this.executionFactory.create(cell, this.isPythonKernelConnection);
+        this.cellExecutions.set(cell, cellExecution);
+        this.queueOfCellsToExecute.push(cellExecution);
+        cellExecution.result.finally(() => this.onCellExecutionCompleted(cellExecution));
```

Comment:
Trying to trace it, looks like this onCellExecutionCompleted might be happening twice? It's on the finally for the result here, and happening when it's executed in executeQueuedCells.